### PR TITLE
Expose diffusion profile in ShaderGraph

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR
Add the ability to expose the diffusion profile asset in the ShaderGraph backboard:
image

### ShaderGraph changes

Note that in order to add this, I had to change the code that registers the blackboard properties to a generic code:
image
As a result of this change, the order of the exposed parameters is now Alphabetical (see screenshot above) compared to the previous one:
image

I also needed to add an API to do the UI of the diffusion profile in the Settings: there is now an interface `IShaderPropertyDrawer` that allows you to implement custom UI code for a certain property type. This generic pattern could apply to all other ShaderGraph properties but it's outside of the scope of this PR.

### Old default exposed diffusion profile

By default when you switched a lit shaderGraph to Transmission or SSS, the old system was exposing a diffusion profile in the material (only available when the drawer was visible in the graph):
image
image

I didn't remove this system because it would break existing ShaderGraph and this means that when you expose one diffusion profile in the Blackboard but don't connect it to the diffusion profile master node input, then you'll have two diffusion profiles in the material UI:
image

What do you think about it?

---
### Testing status

**Manual Tests**: What did you do?
I Manually tested the different combination of the diffusion profile:
- Assign diffusion profile directly from the drawer
- Assign diffusion profile from the automatically exposed profile in the inspector
- Create a diffusion profile parameter and assign it a value in the node settings tab
- Create a diffusion profile parameter and assign it in the material (leave it null in the ShaderGraph)
- Math operations with diffusion profiles:
image
- And the same thing but assigned from the material

I also did some very simple migration tests (the HDRP test project) but I think it'd be good to test it a bit more.

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/902-Graphics/tree/HDRP%252Ffeature%252Fexpose-diffusion-profile-shadergraph

### CODE REVIEW

**Comment by JMargevics** [comment]
Hello! My 2 eurocents:
 
Tested:
- Exporting shader graph with exposed diff. profiles as a package and importing back. No issues ✅
- Assigning default and created profiles. No issues ✅
- Connecting exposed property, disconnecting to see if preview updates correctly. No issues ✅
- Clicking Fix Now under exposed profile in the material. No issues ✅

Minor issues:
6KQ75RyQLe
Why it's called Diffusion Profile (1) when added to graph? Like it's a copy or smth.

image
If you connect exposed profile, it will always use the default green profile for preview, making it pretty unusable (even less than before). Could there be a workaround except editing the default profile asset?

**Comment by alelievr** [comment]
There is a (1) after the name of the parameter is because it's an integer (it's the number of vector components). It does the same thing with vector1 for example.
And you can change the default value of the exposed diffusion profile in the node settings: 
image
Like the other properties of the blackboard (the settings have moved from blackboard to here since the target refactor I think)

**Review by JMargevics — APPROVED** [review]
Thank you very much for explanations! TIL about default diffusion profile in the SG settings. 
I don't have any other comments, everything looks good ✅

**Review by elizabeth-legros — APPROVED** [review]

**Review by cdxntchou — CHANGES_REQUESTED** [review]
Overall I love this changes!  A few things to address in the comments.

In response to your question about the automatic display of the DiffusionProfile property in the Material inspector:
Yeah, it's weird, but I understand fixing it would be tricky.

I think if we wanted to fix it properly we would have to let the Target modify the graph on upgrade to add the property with a matching reference name and hook it up.

Another possibility is to keep it internal in the Target, and just have an advanced setting that Exposes/Hides the auto-created DiffusionProfile property?   If old graphs had this setting set to Expose, and new graphs had it set to Hide, it would mostly do what we want. and be a lot cheaper to implement.

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs
Comment (alelievr) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs:205]: Yeah, I think adding a priority property to the `ShaderInput` class is a good way to give control about order while keeping the "extensibility" of the system.

I'll leave some padding between each item so we can still insert new input types between exiting ones

File: com.unity.shadergraph/Editor/Data/Graphs/AbstractShaderProperty.cs
Comment (alelievr) [reply com.unity.shadergraph/Editor/Data/Graphs/AbstractShaderProperty.cs:37]: It's a leftover from another work, I'll remove it

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/DiffusionProfileShaderProperty.cs
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/DiffusionProfileShaderProperty.cs]: I removed the `SerializableDiffusionProfile` class :)

**Review by JulienIgnace-Unity — APPROVED** [review]

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs
Comment (alelievr) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs]: I finally decided to add an attribute to re-order the shader inputs by type + you can rename them:
image

Here's the new order
image

Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs]: Floats make excellent sort keys, as you can (almost) always insert in between any two keys.

Comment (alelievr) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs]: Yep, I changed the prototype to take a float instead

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs
Comment (cdxntchou) [inline com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs:210]: May be worth using Sort() so you can sort by priority first, then alphabetical for non-priority items.  Would also let us remove most of the priority attribute tags if we don't care about their ordering.

**Review by cdxntchou — CHANGES_REQUESTED** [review]

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs
Comment (alelievr) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs:210]: I thought Sort was a bit too much, but if you want I can do something like this:

```CSharp
            shaderInputTypes.Sort((s1, s2) => {
                var info1 = Attribute.GetCustomAttribute(t, typeof(BlackboardInputInfo)) as BlackboardInputInfo;
                var info2 = Attribute.GetCustomAttribute(t, typeof(BlackboardInputInfo)) as BlackboardInputInfo;
            
                if (info1 == null && info2 == null)
                    return s1.Name.CompareTo(s2.Name);
                else if (info1 != null && info2 == null)
                    return 1;
                else if (info1 == null && info2 != null)
                    return -1;
                else // Both info1 and info2 are not null
                {
                    if (info1.priority == info2.priority)
                        return (info1.name ?? s1.Name).CompareTo(info2.name ?? s2.Name);
                    else
                        return info1.priority.CompareTo(info2.priority);
                }
            });
```

**Comment by sebastienlagarde** [comment]
So Yamato is ready (error come from master):
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/HDRP%252Ffeature%252Fexpose-diffusion-profile-shadergraph/.yamato%252F_abv.yml%2523all_project_ci_CUSTOM-REVISION/3634984/job/pipeline

@cdxntchou  (cc @alelievr ) waiting for your approval to merge it. THanks

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs:210]: yeah that looks great!

Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs:210]: I noticed we have an issue with another PR in flight here (https://github.com/Unity-Technologies/Graphics/pull/381) -- where the Texture2DArrayShaderProperty was subclassed to define special behavior for Lightmap properties..  but we don't want this property type to show up in the menu.

I was thinking: if we make the BlackboardInputInfo attribute required to show up in the menu here, that would let us easily hide internal types by just not tagging them.

And we could use to filter out keywords as well.

**Review by cdxntchou — CHANGES_REQUESTED** [review]

File: com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs
Comment (alelievr) [reply com.unity.shadergraph/Editor/Drawing/Blackboard/BlackboardProvider.cs]: done

**Review by sebastienlagarde — APPROVED** [review]

File: com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/DiffusionProfileShaderProperty.cs
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/ShaderGraph/DiffusionProfileShaderProperty.cs:48]: note: We should move the function asFloat from HDShadowUtils, can be done in another PR

**Review by cdxntchou — APPROVED** [review]
Looks good!

**Pull request merged**
